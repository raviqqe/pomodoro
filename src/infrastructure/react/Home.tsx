import React from "react";
import { PulseLoader } from "react-spinners";
import styled from "styled-components";
import { SignOut } from "./SignOut";

const Container = styled.div`
  display: flex;
  flex-direction: column;
  height: 100vh;
  width: 85ex;
  max-width: 100%;
  margin: auto;

  > :first-child {
    flex: 1;
  }
`;

const LoaderContainer = styled.div`
  display: flex;
  justify-content: center;
  align-items: center;
`;

const SignOutContainer = styled.div`
  position: fixed;
  top: 0.5em;
  right: 0.5em;
`;

export interface IProps {
  signOut: () => Promise<void>;
}

export const Home = ({ signOut }: IProps) => {
  return (
    <Container>
      <LoaderContainer>
        <PulseLoader color="white" />
      </LoaderContainer>
      <SignOutContainer>
        <SignOut signOut={signOut} />
      </SignOutContainer>
    </Container>
  );
};
