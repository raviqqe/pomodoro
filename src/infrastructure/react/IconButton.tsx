import React, { AriaAttributes, ReactNode } from "react";
import styled from "styled-components";

const Button = styled.div`
  text-shadow: 0rem 0.2rem 0.2rem rgba(0, 0, 0, 0.2);
  color: grey;
  cursor: pointer;
  font-size: 1.5em;
  display: flex;
`;

interface IProps extends AriaAttributes {
  children: ReactNode;
  onClick: () => void;
}

export const IconButton = ({ children, onClick, ...restProps }: IProps) => (
  <Button onClick={onClick} {...restProps}>
    {children}
  </Button>
);
