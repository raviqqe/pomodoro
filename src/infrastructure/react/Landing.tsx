import { GoMarkGithub } from "react-icons/go";
import React from "react";
import styled from "styled-components";
import { SignIn } from "./SignIn";

const Container = styled.div`
  height: 100vh;
  width: 100vw;
  display: flex;
  flex-direction: column;
  justify-content: center;
  align-items: center;

  > * {
    margin: 2rem;
  }
`;

const Title = styled.div`
  color: steelblue;
  font-family: "Chelsea Market", sans-serif;
  font-size: 4em;
  font-weight: bold;
  letter-spacing: -0.35ex;
  text-shadow: 0.4em 0.4em 0.4em rgba(0, 0, 0, 0.1);
`;

const GitHubLink = styled.a`
  font-size: 2.5rem;
  margin: 1.5rem;
  color: #222;
  display: block;
  position: fixed;
  bottom: 0em;
  right: 0em;
  line-height: 0ex;
`;

export interface IProps {
  repositoryURL: string;
  signIn: () => void;
}

export const Landing = ({ repositoryURL, signIn }: IProps) => (
  <Container>
    <Title>P🍅m🍅d🍅r🍅</Title>
    <SignIn signIn={signIn} />
    <GitHubLink href={repositoryURL} target="_blank">
      <GoMarkGithub />
    </GitHubLink>
  </Container>
);
