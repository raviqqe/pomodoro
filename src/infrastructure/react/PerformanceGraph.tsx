import { last } from "lodash";
import { DateTime } from "luxon";
import {
  CartesianGrid,
  ResponsiveContainer,
  XAxis,
  YAxis,
  Label,
  Bar,
  BarChart,
} from "recharts";
import styled from "styled-components";
import { IPerformanceGraph } from "../../application/performance-graph";
import { DateSerializer } from "../../domain/date-serializer";
import { white } from "./style/colors";

const Container = styled.div`
  width: 80vw;
  height: 80vh;
`;

const Message = styled.div`
  color: ${white};
  font-size: 1.5em;
`;

export interface IProps {
  performanceGraph: IPerformanceGraph;
}

export const PerformanceGraph = ({
  performanceGraph: { data },
}: IProps): JSX.Element =>
  data.length === 0 ? (
    <Message>No performance graph to show yet!</Message>
  ) : (
    <Container>
      <ResponsiveContainer>
        <BarChart data={data}>
          <CartesianGrid fill="white" stroke="grey" strokeDasharray="3 3" />
          <XAxis
            dataKey="date"
            stroke="grey"
            tickFormatter={(date: string): string => {
              const days: number = DateTime.fromJSDate(
                DateSerializer.deserialize(
                  (last(data) as { date: string }).date
                )
              ).diff(
                DateTime.fromJSDate(DateSerializer.deserialize(date)),
                "days"
              ).days;

              return days === 0 ? "Today" : `${days} days ago`;
            }}
            tickMargin={10}
          />
          <YAxis allowDecimals={false} stroke="grey" tickMargin={5}>
            <Label
              angle={-90}
              position="insideLeft"
              style={{ fill: "grey" }}
              value="Pomodoros"
            />
          </YAxis>
          <Bar dataKey="pomodoros" fill="salmon" />
        </BarChart>
      </ResponsiveContainer>
    </Container>
  );
