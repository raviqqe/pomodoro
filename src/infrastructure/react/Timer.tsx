import styled from "styled-components";
import { PomodoroTimerState } from "../../application/pomodoro-timer-state";
import { TextButton } from "./TextButton";
import { white } from "./style/colors";

const Container = styled.div`
  display: flex;
  flex-direction: column;
  align-items: center;

  > * {
    margin: 1rem;
  }
`;

const State = styled.div`
  font-size: 5em;
`;

const Time = styled.div`
  color: ${white};
  font-size: 10em;
`;

const Minutes = styled.span``;

const Seconds = styled.span`
  font-size: 0.6em;
  margin-left: 0.2ex;
`;

export interface IProps {
  seconds: number;
  startTimer: () => void;
  state: PomodoroTimerState;
  stopTimer: () => void;
  stopped: boolean;
}

export const Timer = ({
  seconds,
  startTimer,
  state,
  stopTimer,
  stopped,
}: IProps): JSX.Element => (
  <Container>
    {stopped ? (
      <>
        <State>
          {state === PomodoroTimerState.Pomodoro
            ? "🍅"
            : state === PomodoroTimerState.ShortBreak
            ? "🛌"
            : "🛌🛌"}
        </State>
        <TextButton onClick={startTimer}>Start</TextButton>
      </>
    ) : (
      <>
        <Time>
          <Minutes>{Math.floor(seconds / 60)}</Minutes>
          <Seconds>{seconds % 60}</Seconds>
        </Time>
        <TextButton onClick={stopTimer} secondary={true}>
          Stop
        </TextButton>
      </>
    )}
  </Container>
);
