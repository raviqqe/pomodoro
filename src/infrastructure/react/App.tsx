import { Omit } from "lodash";
import { PulseLoader } from "react-spinners";
import { useAsync } from "react-use";
import React, { useState } from "react";
import styled from "styled-components";
import { IInitialState } from "../../application/application-initializer";
import { IProps as ILandingProps, Landing } from "./Landing";
import { Home, IProps as IHomeProps } from "./Home";

const LoaderContainer = styled.div`
  display: flex;
  justify-content: center;
  align-items: center;
  height: 100vh;
  width: 100vw;
`;

interface IProps extends Omit<IHomeProps, "signOut">, ILandingProps {
  initialize: () => Promise<IInitialState>;
  signIn: () => Promise<boolean>;
  signOut: () => Promise<boolean>;
}

export const App = ({
  initialize,
  repositoryURL,
  signIn,
  signOut,
  ...props
}: IProps) => {
  const [signedIn, setSignedIn] = useState<boolean | null>(null);

  useAsync(async () => {
    const { signedIn } = await initialize();
    setSignedIn(signedIn);
  }, []);

  if (signedIn === null) {
    return (
      <LoaderContainer>
        <PulseLoader color="white" />
      </LoaderContainer>
    );
  } else if (!signedIn) {
    return (
      <Landing
        repositoryURL={repositoryURL}
        signIn={async () => setSignedIn(await signIn())}
      />
    );
  }

  return <Home {...props} signOut={async () => setSignedIn(await signOut())} />;
};
