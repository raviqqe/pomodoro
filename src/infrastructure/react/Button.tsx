import { FC, HTMLAttributes, HTMLProps, ReactNode } from "react";
import styled from "styled-components";
import { boxShadow } from "./style";

interface IProps
  extends HTMLProps<HTMLButtonElement>,
    HTMLAttributes<HTMLButtonElement> {
  secondary?: boolean;
  children?: ReactNode;
}

export const Button: FC<IProps> = styled.button<{ secondary?: boolean }>`
  ${boxShadow};
  background: ${({ secondary }) => (secondary ? "darkgrey" : "salmon")}
  display: flex;
  justify-content: center;
  align-items: center;
  font-family: inherit;
  color: white;
  border: none;
  cursor: pointer;
  flex-shrink: 0;
` as FC<any>;
