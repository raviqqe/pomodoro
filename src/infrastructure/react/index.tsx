import { render } from "react-dom";
import React from "react";
import { ApplicationInitializer } from "../../application/application-initializer";
import { SignInManager } from "../../application/sign-in-manager";
import { SignOutManager } from "../../application/sign-out-manager";
import { PerformanceGraphViewer } from "../../application/performance-graph-viewer";
import { PomodoroTimerStopper } from "../../application/pomodoro-timer-stopper";
import { PomodoroTimerStarter } from "../../application/pomodoro-timer-starter";
import { PomodoroTimerState } from "../../application/pomodoro-timer-state";
import { IRenderer, IPomodoroTimer } from "../renderer";
import { IPerformanceGraph } from "../../application/performance-graph";
import { App, IProps as IAppProps } from "./App";
import { GlobalStyle } from "./style";

interface IPresenter {
  setRenderer(renderer: IRenderer): void;
}

interface IProps
  extends Pick<IAppProps, "performanceGraph" | "signedIn" | "timer"> {}

export class ReactRenderer implements IRenderer {
  private props: IProps = {
    performanceGraph: { data: [] },
    signedIn: null,
    timer: { seconds: 0, state: PomodoroTimerState.Pomodoro, stopped: true },
  };

  constructor(
    private readonly element: HTMLElement,
    presenters: IPresenter[],
    private readonly applicationInitializer: ApplicationInitializer,
    private readonly performanceGraphViewer: PerformanceGraphViewer,
    private readonly pomodoroTimerStarter: PomodoroTimerStarter,
    private readonly pomodoroTimerStopper: PomodoroTimerStopper,
    private readonly signInManager: SignInManager,
    private readonly signOutManager: SignOutManager,
    private readonly repositoryURL: string
  ) {
    for (const presenter of presenters) {
      presenter.setRenderer(this);
    }
  }

  public render(): void {
    this.renderProps({});
  }

  public renderPerformanceGraph(performanceGraph: IPerformanceGraph): void {
    this.renderProps({ performanceGraph });
  }

  public renderSignedIn(signedIn: boolean): void {
    this.renderProps({ signedIn });
  }

  public renderTimer(timer: IPomodoroTimer): void {
    this.renderProps({ timer });
  }

  private renderProps(props: Partial<IProps>): void {
    this.props = { ...this.props, ...props };

    render(
      <>
        <App
          {...this.props}
          initialize={() => this.applicationInitializer.initialize()}
          repositoryURL={this.repositoryURL}
          signIn={() => this.signInManager.signIn()}
          signOut={() => this.signOutManager.signOut()}
          startTimer={() => this.pomodoroTimerStarter.start()}
          stopTimer={() => this.pomodoroTimerStopper.stop()}
          viewGraph={() => this.performanceGraphViewer.viewGraph()}
        />
        <GlobalStyle />
      </>,
      this.element
    );
  }
}
