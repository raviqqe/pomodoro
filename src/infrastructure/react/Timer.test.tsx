import { render } from "@testing-library/react";
import { expect, it } from "vitest";
import { PomodoroTimerState } from "../../application/pomodoro-timer-state";
import { Timer, IProps } from "./Timer";

const commonProps: Omit<IProps, "state" | "stopped"> = {
  seconds: 42,
  startTimer: async () => {},
  stopTimer: async () => {},
};

it("renders while running", () => {
  expect(
    render(
      <Timer
        {...commonProps}
        state={PomodoroTimerState.Pomodoro}
        stopped={false}
      />
    ).container.firstChild
  ).toMatchSnapshot();
});

it("renders with pomodoro state", () => {
  expect(
    render(
      <Timer
        {...commonProps}
        state={PomodoroTimerState.Pomodoro}
        stopped={true}
      />
    ).container.firstChild
  ).toMatchSnapshot();
});

it("renders with short break state", () => {
  expect(
    render(
      <Timer
        {...commonProps}
        state={PomodoroTimerState.ShortBreak}
        stopped={true}
      />
    ).container.firstChild
  ).toMatchSnapshot();
});

it("renders with long break state", () => {
  expect(
    render(
      <Timer
        {...commonProps}
        state={PomodoroTimerState.LongBreak}
        stopped={true}
      />
    ).container.firstChild
  ).toMatchSnapshot();
});
