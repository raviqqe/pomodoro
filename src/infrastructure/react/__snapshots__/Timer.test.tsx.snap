// Vitest Snapshot v1

exports[`renders while running 1`] = `
<div
  class="sc-dkrFOg IsHIT"
>
  <div
    class="sc-eDvSVe dXemcg"
  >
    <span
      class="sc-jSUZER"
    >
      0
    </span>
    <span
      class="sc-gKPRtg PXMhH"
    >
      42
    </span>
  </div>
  <button
    class="sc-bcXHqe sc-gswNZR fmEPPj gCUrWJ"
  >
    Stop
  </button>
</div>
`;

exports[`renders with long break state 1`] = `
<div
  class="sc-dkrFOg IsHIT"
>
  <div
    class="sc-hLBbgP blDmMu"
  >
    🛌🛌
  </div>
  <button
    class="sc-bcXHqe sc-gswNZR eTvdtE gCUrWJ"
  >
    Start
  </button>
</div>
`;

exports[`renders with pomodoro state 1`] = `
<div
  class="sc-dkrFOg IsHIT"
>
  <div
    class="sc-hLBbgP blDmMu"
  >
    🍅
  </div>
  <button
    class="sc-bcXHqe sc-gswNZR eTvdtE gCUrWJ"
  >
    Start
  </button>
</div>
`;

exports[`renders with short break state 1`] = `
<div
  class="sc-dkrFOg IsHIT"
>
  <div
    class="sc-hLBbgP blDmMu"
  >
    🛌
  </div>
  <button
    class="sc-bcXHqe sc-gswNZR eTvdtE gCUrWJ"
  >
    Start
  </button>
</div>
`;
